File: challenges/interviewbit/programming/arrays/set-matrix-zeros/set-matrix-zeros.py
# pylint: disable=missing-docstring

def haveZeroInFirstRow(matrix):
  for i in xrange(len(matrix[0])):
    if matrix[0][i] == 0:
      return True
  return False


def setAllZeroInFirstRow(matrix):
  for i in xrange(len(matrix[0])):
    matrix[0][i] = 0
  return matrix


def haveZeroInFirstColumn(matrix):
  for i in xrange(len(matrix)):
    if matrix[i][0] == 0:
      return True
  return False


def setAllZeroInFirstColumn(matrix):
  for i in xrange(len(matrix)):
    matrix[i][0] = 0
  return matrix

def setInfoInFirstRowAndFirstColumn(matrix):
  for row in xrange(1, len(matrix)):
    for col in xrange(1, len(matrix[row])):
      if matrix[row][col] == 0:
        matrix[0][col] = 0
        matrix[row][0] = 0
  return matrix

def setMatrixZeros(matrix):
  zeroInFirstRow = haveZeroInFirstRow(matrix)
  zeroInFirstColumn = haveZeroInFirstColumn(matrix)
  matrix = setInfoInFirstRowAndFirstColumn(matrix)

  for row in xrange(1, len(matrix)):
    for col in xrange(1, len(matrix[row])):
      matrix[row][col] = 1 if matrix[0][col] == 1 and matrix[row][0] == 1 else 0

  if zeroInFirstRow:
    matrix = setAllZeroInFirstRow(matrix)

  if zeroInFirstColumn:
    matrix = setAllZeroInFirstColumn(matrix)

  return matrix

File: packages/interviewbit/programming/arrays/set-matrix-zeros/set-matrix-zeros.ts
/**
 * Interviewbit - SetMatrixZeros
 */

function haveZeroInFirstRow(matrix: number[][]): boolean {
  // tslint:disable-next-line
  for (let i: number = 0; i < matrix[0].length; i = i + 1) {
    if (matrix[0][i] === 0) {
      return true;
    }
  }

  return false;
}

function setAllZeroInFirstRow(matrix: number[][]): number[][] {
  for (let i: number = 0; i < matrix[0].length; i = i + 1) {
    matrix[0][i] = 0;
  }

  return matrix;
}

function haveZeroInFirstColumn(matrix: number[][]): boolean {
  // tslint:disable-next-line
  for (let i: number = 0; i < matrix.length; i = i + 1) {
    if (matrix[i][0] === 0) {
      return true;
    }
  }

  return false;
}

function setAllZeroInFirstColumn(matrix: number[][]): number[][] {
  // tslint:disable-next-line
  for (let i: number = 0; i < matrix.length; i = i + 1) {
    matrix[i][0] = 0;
  }

  return matrix;
}

function setInfoInFirstRowAndFirstColumn(matrix: number[][]): number[][] {
  for (let row: number = 1; row < matrix.length; row = row + 1) {
    for (let col: number = 1; col < matrix[row].length; col = col + 1) {
      if (matrix[row][col] === 0) {
        matrix[0][col] = 0;
        matrix[row][0] = 0;
      }
    }
  }

  return matrix;
}

function setMatrixZeros(matrix: number[][]): number[][] {
  const zeroInFirstRow: boolean = haveZeroInFirstRow(matrix);
  const zeroInFirstColumn: boolean = haveZeroInFirstColumn(matrix);

  // tslint:disable-next-line
  matrix = setInfoInFirstRowAndFirstColumn(matrix);

  for (let row: number = 1; row < matrix.length; row = row + 1) {
    for (let col: number = 1; col < matrix[row].length; col = col + 1) {
      matrix[row][col] = matrix[0][col] === 1 && matrix[row][0] === 1 ? 1 : 0;
    }
  }

  if (zeroInFirstRow) {
    // tslint:disable-next-line
    matrix = setAllZeroInFirstRow(matrix);
  }

  if (zeroInFirstColumn) {
    // tslint:disable-next-line
    matrix = setAllZeroInFirstColumn(matrix);
  }

  return matrix;
}

export { setMatrixZeros };
